Create a new method to generalize the offensive ability usages

JasonKaz opened the pull request.

JasonKaz committed:
Create a new method to generalize the offensive ability usages

mouseas commented on autoPlay.js:
```diff
@@ -297,37 +297,16 @@ function useClusterBombIfRelevant() {
-			return;
-		}
-		
-		//Check lane has monsters to burn
```
While a generalized method would be helpful, you can't but out the logic in each of the methods.

Also, I think naming the method `useOffensiveAbilityIfAvailable` would be better, then just do the ability-specific logic before calling it, and you should be good.

JasonKaz committed:
Add ability usage condition map

JasonKaz committed:
Add conditions for MORALE_BOOSTER

JasonKaz commented:
So with these new changes it allows us to have a configurable list of conditions for using abilities. I didn't move the medic or charm abilities because they seemed very specific and weren't easy to generalize.

mouseas commented:
I need ~~2 people~~ one more person to independently test this for errors and functionality. (Thanks @furtivespy)

furtivespy commented:
I haven't tested, but I've reviewed the code and the logic looks sound. Here are my questions/comments:
- Get rid of the changed useAbilityIfRelevant() functions. just call the generic from main.
- Lucky charms can be done - just null on all except enemy count 2+ (it's only not used on boss).
- Should abilityConditions be part of a global config item? Either way should probably stick with the standard of ALL_CAPS for global variables.

JasonKaz committed:
Update to latest

-Updated to latest version of file
-Moved ABILITY_CONDITIONS to top
-Removed methods to call a single ability, now the main method calls the generic one
-Added ITEM_CONDITIONS and useItemIfAvailable() for CRIPPLE_SPAWNER and GOLD_RAIN

JasonKaz commented:
All of @furtivespy 's changes were made, plus others.

We really should get these changes in sooner rather than later. It makes adding new ability/item using very trivial.

JasonKaz commented:
We really should get these changes in sooner rather than later. It makes adding new ability/item using very trivial.

mouseas commented on autoPlay.js:
```diff
@@ -1,458 +1,510 @@
-	var enemyTypePriority = [
-		ENEMY_TYPE.TREASURE, 
-		ENEMY_TYPE.BOSS, 
+    // We can overlook spawners if all spawners are 40% hp or higher and a creep is under 10% hp
```
Please revert the whitespace changes. The code is formatted with tabs, and you've switched a bunch of it to spaces.

JasonKaz committed:
Spaces to tabs

JasonKaz commented:
Should be fully converted now.

mouseas commented:
@JasonKaz Great, now we just need someone to test it.

win0na commented:
When this is ready, inform me and I will fix the various merge conflicts introduced by the filename change.

JasonKaz commented:
I don't suppose just me testing it is enough?

win0na commented:
As indicated by this comment (https://github.com/mouseas/steamSummerMinigame/pull/42#issuecomment-111672330), it would be beneficial to have at least one more person testing it to ensure it works properly.

This macro explains it perfectly.

![descarga1](https://cloud.githubusercontent.com/assets/4269491/8145831/9e8c835c-11de-11e5-9995-3528d4de16d9.jpg)

mouseas commented:
This PR is too old to merge in, and code duplication has already been reduced sufficiently for the areas in question. I'm closing this PR.
